R/with-qenv.R

```diff
@@ -0,0 +1,411 @@
+#' qenv refactor prototype
+#'
+#' Simple to use environment with history tracking.
+#'
+#' @details
+#' Create a `quenv` object, which is an environment, and execute code inside.
+#' Code can be supplied as expressions, literal character vectors, as well as name-bound character vectors.
+#' External values can be injected into the code with the ellipsis.
+#'
+#' `quenv` creates `quenv` object.
+#' `with` acts in `quenv` object.
+#' `within` creates and modifies a (deep) copy of `quenv` object.
+#' `get_code` returns list of function calls or a data.frame with code and the conditions it raised.
+#' `get_conditions` returns list of condition messages (character strings).
+#'
+#' @param data,x (`quenv`)
+#' @param expr (`language`) simple or compound expression to evaluate in `data`
+#' @param text (`character`) character vector of expressions to evaluate in `data`
+#' @param ... `name:value` pairs to inject values into `expr`
+#'
+#' @return
+#' `quenv` returns a `quenv` object. `with` returns NULL invisibly. `within` returns a modified deep copy of `data`.
+#'
+#' @name quenv
+#'
+#' @examples
+#'
+#' q <- quenv()
+#'
+#' # execute code
+#' with(q, {
+#'   i <- iris
+#'   m <- mtcars
+#' })
+#' q
+#'
+#' # supply code as strings
+#' q <- quenv()
+#' with(q, text = "c <- cars")
+#' code_as_text <- "w <- warpbreaks"
+#' with(q, text = code_as_text)
+#'
+#' # error messages are stored
+#' try(
+#'   with(q, {
+#'     subset(i, Species == species) # raises error and stops evaluation
+#'     ms <- subset(m, cyl == 4) # not evaluated
+#'   })
+#' )
+#' q
+#'
+#' # warnings and messages are also stored
+#' with(q, {
+#'   warning("this is a warning")
+#' })
+#' with(q, {
+#'   message("this is a message")
+#' })
+#' q
+#'
+#' # access variables and environment history
+#' q$m
+#' get_code(q)
+#' get_conditions(q, "error")
+#'
+#' # inject values into code
+#' q <- quenv()
+#' with(q, i <- iris)
+#' with(q, print(dim(subset(i, Species == "virginica"))))
+#' try(
+#'   with(q, print(dim(subset(i, Species == species)))) # fails
+#' )
+#' with(q, print(dim(subset(i, Species == species))), species = "versicolor")
+#' species_external <- "versicolor"
+#' with(q, print(dim(subset(i, Species == species))), species = species_external)
+#'
+#' # execute code in copy of `quenv` and return modified copy
+#' q <- quenv()
+#' with(q, i <- iris)
+#' qq <- within(q, m <- mtcars)
+#'
+
+#' @rdname quenv
+#' @export
+quenv <- function() {
+  ans <- new.env()
+  attr(ans, "code") <- list()
+  attr(ans, "errors") <- list()
+  attr(ans, "warnings") <- list()
+  attr(ans, "messages") <- list()
+  class(ans) <- c("quenv", class(ans))
+  ans
+}
```

Comment:
```suggestion
quenv <- function() {
  ans <- new.env()
  attr(ans, "code") <- list()
  attr(ans, "errors") <- list()
  attr(ans, "warnings") <- list()
  attr(ans, "messages") <- list()
  class(ans) <- c("quenv", class(ans))
  ans
}
```

Smart move with qenv being an environment. Not sure yet what is better:
1. qenv as S3 environment and protecting class by overwriting `$<-.qenv`, `[[<-.qenv` but potentially also other methods (which? `eval`?)
2. qenv as S4 and writing own `$`, `[[` methods to get objects.

I like `qenv` as environment, I think inheritance and reusing the class in other packages should be the key for a decision.